core/src/main/scala/org/apache/spark/sql/execution/datasources/StoreDataSourceStrategy.scala

```diff
@@ -42,25 +42,10 @@ private[sql] object StoreDataSourceStrategy extends Strategy{
         filters,
         t.numPartitions,
         t.partitionColumns,
-        (a, f) => toCatalystRDD(l, a, t.buildScan(a.map(_.name).toArray, f))) :: Nil
-
+        (a, f) => t.buildUnsafeScan(a.map(_.name).toArray, f)) :: Nil
```

Comment:
Also add handling for just PrunedUnsafeFilteredScan in addition to PartitionedDataSourceScan. For testing I am using something like this in addition to these changes:

```scala
    case PhysicalOperation(projects, filters,
        l@LogicalRelation(t: PrunedUnsafeFilteredScan, _, _)) =>
      pruneFilterProject(
        l,
        projects,
        filters,
        0,
        Seq.empty[String],
        (a, f) => t.buildUnsafeScan(a.map(_.name).toArray, f)) :: Nil
```

then in pruneFilterProjectRaw add handling for numPartition == 0 like:

```scala
      val scan = if (numPartition > 0) {
        execution.PartitionedPhysicalRDD.createFromDataSource(
          projects.map(_.toAttribute),
          numPartition,
          joinedCols,
          scanBuilder(requestedColumns, candidatePredicates, pushedFilters),
          relation.relation)
      } else {
        execution.DataSourceScanExec.create(
          projects.map(_.toAttribute),
          scanBuilder(requestedColumns, candidatePredicates, pushedFilters),
          relation.relation, metadata, relation.metastoreTableIdentifier)
      }
```

I will attach my full patch to this file to JIRA.